server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -278,6 +279,28 @@ public void getSynonymRule(String synonymSetId, String synonymRuleId, ActionList
         );
     }
 
+    public void deleteSynonymRule(
+        String synonymSetId,
+        String synonymRuleId,
+        ActionListener<SynonymsReloadResult<AcknowledgedResponse>> listener
+    ) {
+        checkSynonymSetExists(
+            synonymSetId,
+            listener.delegateFailure(
+                (l1, obj) -> client.prepareDelete(SYNONYMS_ALIAS_NAME, internalSynonymRuleId(synonymSetId, synonymRuleId))
+                    .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                    .execute(new DelegatingIndexNotFoundActionListener<>(synonymSetId, l1, (l2, deleteResponse) -> {
+                        if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+                            l2.onFailure(new ResourceNotFoundException("synonym rule [" + synonymRuleId + "] not found"));
```

Comment:
Question: Should this error message display only the synonym rule ID? Or would it be helpful to include the synonym set ID as well for troubleshooting?

Reply:
I wonder if need at all to have `checkSynonymSetExists` here? Isn't it enough just to issue DELETE request and if it returns "NOT_FOUND" return `ResourceNotFoundException`?  
By this we will avoid extra search request. WDYT?

Reply:
> would it be helpful to include the synonym set ID as well for troubleshooting?

Yes! Adding it. Thanks!

> I wonder if need at all to have checkSynonymSetExists here?

I think it's needed, as otherwise we don't know if NOT_FOUND refers to not finding the synonym rule or the entire synonym set.

Reply:
> I think it's needed, as otherwise we don't know if NOT_FOUND refers to not finding the synonym rule or the entire synonym set.

@carlosdelest Do you think we can have an error message like this: "synonym_set or synonym_rule not found"? It concerns me that we need to do an extra search request just for an error message.

Also, there is a tiny possibility that `synonym_set` gets deleted after `checkSynonymSetExists` is executed and before synonym rule delete request.

Reply:
> Do you think we can have an error message like this: "synonym_set or synonym_rule not found"?

I find it less explicit, though we can definitely do that if needed.

> It concerns me that we need to do an extra search request just for an error message.

Given this is an admin like operation (not impacting search) and that it shouldn't be that frequent, do you think we need to optimize for an additional search request vs providing a more helpful message?

If so, how about reversing the check, and do the additional check on the synonym set when we have a NOT FOUND, so we can provide a more helpful message?

> Also, there is a tiny possibility that synonym_set gets deleted after checkSynonymSetExists is executed and before synonym rule delete request.

Yes, there is always that risk in all operations. We need to deal with concurrency on a separate task I think.

Reply:
> how about reversing the check, and do the additional check on the synonym set when we have a NOT FOUND, so we can provide a more helpful message?

I like the idea, I think it is a good compromise.

Reply:
I've implemented that, LMKWYT!